### DISCUSSION TITLE
Letter Spacing Issue/Feature Request

### DISCUSSION THREAD

**Discussion by samcx** [ghostty-org/ghostty#3842] [Feature Requests, Ideas]
I'm seeing slightly (maybe 1 pixel difference?) different letter spacing between 👻 and iTerm. Not sure if this is something custom with iTerm (don't believe I've adjusted it there). It would be nice if we could adjust this in 👻.
(Top is 👻, it is longer in width because the letter spacing)

![CleanShot 2024-12-28 at 19 09 34@2x](https://github.com/user-attachments/assets/9512d78d-a0ec-4322-8b60-4482c91cd9a9)

**Comment by jmckillican** [comment]
I personally prefer the spacing of ghostty when comparing those two.  some of the letters/characters appear too close to each other at times.   you can see it in words like 'biome.json' where there's barely any spacing between 'm' and 'e'.

**Comment by mitchellh** [comment]
Ghostty's spacing does look better to me too. But, just to be sure we're doing things right, what font is this?

**Reply by samcx** [reply]
@mitchellh This is Geist Mono → https://vercel.com/font

**Comment by neel04** [comment]
Having the same problem with Jetbrains Mono :( (👻 on top vs WezTerm):
![image](https://github.com/user-attachments/assets/678b51fb-8894-45f3-9e2b-9038bd88ab93)
![image](https://github.com/user-attachments/assets/a8da1930-471c-4b96-bb3c-3769005dace4)

Its pretty noticeable and very annoying

**Reply by jmckillican** [reply]
what exactly it the problem tho?   I struggle to see how it's "very" annoying unless you always keep wezterm terminal right on top of your ghostty one and mirroring the exact same output.

**Reply by neel04** [reply]
Its definitely a lot easier to see when you're just working with a whole screen of code.

When you're used to something, all these differences just add up and feel *very* foreign. 

I recognize that this is a really minor nitpick, but even minor variations in font spacing, line widths etc. immediately jump out to someone used to staring at that font for hours. 

Its just hard to switch to something new when its aesthetically wanting. I'll be sticking to WezTerm for now, but I would completely understand if the devs don't want to work towards fixing this issue...

**Reply by rhodes-b** [reply]
i just saw this, the spacing is the same here, the font thickness is different use font thicken on macos or a variable font (the default ghostty font is not one) and adjust the weight with font-variation as listed here
https://ghostty.org/docs/config/reference#font-variation

**Comment by suciuvlad** [comment]
I have same the _issue_ when comparing it to iTerm. I'm using MesloLGS NF. Overall, it feels different.

**Reply by suciuvlad** [reply]
<img width="988" alt="image" src="https://github.com/user-attachments/assets/56b03e34-7333-4ab1-a460-4b5d95450b69" />

**Comment by dmd** [comment]
Same. This immediately turned me off Ghostty. It just ... looks wrong, when I've been staring at the same font with same kerning for 20+ years of my life.

**Comment by cameronmurphy** [comment]
I've gotten used to JetBrains Mono size 13 with `font-thicken` on but overall I find it less readable than iTerm.

**Comment by crabique** [comment]
Setting these made it look more or less the same as in iTerm2, judging by the proportions of the Apple logo in `neofetch`:
```hcl
adjust-cell-width = "-5%"
adjust-cell-height = "5%"
```

**Comment by klopsknoedel** [comment]
Maybe this behaviour could be related to https://github.com/ghostty-org/ghostty/discussions/4197
I face the exact same issue comparing the rendering on my MacBook's display to an external 1440p monitor, where I have to set `adjust-cell-width = "-7%"` to match the internal display.

MacBook Retina Display
<img width="799" alt="Screenshot 2025-01-03 at 10 09 55" src="https://github.com/user-attachments/assets/a17395f3-7f1e-4b01-aecc-823bcd7a8e10" />


External Monitor
![Screenshot 2025-01-03 at 10 10 16](https://github.com/user-attachments/assets/2137cad0-ec04-4b00-bfa9-594bfd26c9cf)


You can clearly see the distortion comparing the apple logos.

**Reply by mitchellh** [reply]
Note I opened #4504 for this.

**Comment by avatus** [comment]
I'm not sure if its related/similar but it happens with the font glyphs as well. This font is JetBrainsMono Nerd Font and the "up arrow" is out of whack on ghostty (left) compared to warp (right) 
![Screenshot 2025-01-03 at 10 34 02 AM](https://github.com/user-attachments/assets/ddb4e927-fcd7-48ae-9f31-746382b393b6)


same font family, font size
i tried font-thicken and, while it did make it thicker, didnt really change anything else

**Reply by mitchellh** [reply]
That looks like a font fallback issue.

**Reply by avatus** [reply]
Thanks. I'm not quite sure what that means. As in, this font isn't being found and its trying to fallback to some unsupported font? If thats the case, I'll look and see what fonts my other terminals fall back to in order to properly display this glyph. Thanks for the input. Let me know if I misunderstood

**Reply by mitchellh** [reply]
Yeah, it looks like it's choosing a totally different font for that glyph for some reason. There was another discussion this came up too..

**Comment by klaustopher** [comment]
I got the same problem using Fantasque Sans Mono font, on the native screen on the mac it looks okay and I can live with the subtle difference between ghostty and iTerm. But on the external screen (which is my main screen to work on), the spacing is completely outside of anything my eyes can parse :)

external screen (3440x1440):
![Bildschirmfoto 2025-01-06 um 10 00 59](https://github.com/user-attachments/assets/13c6f4ff-b85f-4c07-9dec-74fdcdc0a2f1)

native screen in "more area" mode (1710x1112)
<img width="828" alt="Bildschirmfoto 2025-01-06 um 10 01 42" src="https://github.com/user-attachments/assets/9fbd9e55-afd8-4199-abe5-e21b1538a2f6" />

ghostty is on top, iTerm2 on the bottom

LMK if you need more infos

**Reply by rhodes-b** [reply]
scroll up a bit same issue with a response https://github.com/ghostty-org/ghostty/discussions/3842#discussioncomment-11724177

**Comment by mitchellh** [comment]
Note there's a lot of screenshots here but I think the question I have is mainly: what is correct? For note, I haven't looked at iTerms code and how they calculate their grid metrics. 

We base our spacing off of the font's published font metrics. To the best of my knowledge, this is strictly correct. We don't perform heuristics like other terminals do (I don't know if iTerm does this), such as basing grid with based on the lowercase "m" (since it _tends_ to be wide, some terminals _do in fact_ do this). We defer spacing completely to what the font author tells us it should be. 

Our calculations could, of course, be wrong. If so, I'd like to know. That's an easy fix.

If they're not wrong, then I think the correct answer is to either open an issue with various fonts, or to apply cell width adjustments. I think that's annoying but its in the realm of subjective choice that I don't think we can safely apply to all fonts out there. I think leaning on what font authors advertise is a valid and expected behavior and creating our own spacing heuristics seems long term like a bad idea...

**Reply by neel04** [reply]
https://github.com/ghostty-org/ghostty/discussions/3842#discussioncomment-11722000 seems to fix it for me, so I wonder if its some font spacing calculation issue. I've tried lots of different terminals over the years (the default, iTerm, Wezterm, kitty) and they all render the same font similarly so it seems like a ghostty issue tbh

**Reply by klaustopher** [reply]
Personally, I think that the subtle difference on the normal mac screen is fine. But something is very off with the rendering on external screens, but it seems we are gathering that in #4504.

**Reply by huylv** [reply]
Here's a comparison between Ghostty and https://www.jetbrains.com/lp/mono/. White font is Ghostty. I hope Jetbrains renders it correctly :-)
<img width="513" alt="image" src="https://github.com/user-attachments/assets/f5acbd9b-04fd-4a6c-951e-ab0eb82d458a" />

**Reply by mitchellh** [reply]
To nitpick, Jetbrains isn't rendering anything. That'd be your browser vendor, presumably. Browsers _do_ generally render fonts correctly and they have some of the best font rendering systems out there. But they're also highly variable due to their customization (CSS) and slightly different constraints than a terminal (the pre-computed monospace grid). 

I'm not saying you're wrong, but I still need something a bit more concrete. I suspect browsers have something very specific with regards to how they space graphemes, but I can't find the specification for that. If someone did, that'd be helpful to compare to our approach.

**Comment by S1lander** [comment]
For anyone still looking for a solution ... in my case (using the powerlevel10k theme with the default Meslo Nerd Font) these values get me very close to how it looks in iTerm:
```
# font stuff
font-family = "MesloLGS NF"
font-size = 12
font-thicken = true
adjust-cell-width = "-10%"
```
Top = Ghostty, Bottom = iTerm
![ghostty_font_fix](https://github.com/user-attachments/assets/acb17e54-b8be-4a19-acc1-1993d4b697b0)

**Comment by injust** [comment]
I've been using `adjust-cell-width = -1` to fix the cell width being too large.

After https://github.com/ghostty-org/ghostty/pull/9432, I can remove the adjustment and the cell width looks a lot more correct!
